bin/propolis-server/src/lib/migrate/destination.rs

```diff
@@ -24,105 +27,321 @@ use crate::migrate::probes;
+    // TODO: We need to make sure the src_addr is a valid target
+    let src_migrate_url = format!(
+        "ws://{}/instance/migrate/{}/start",
+        migrate_info.src_addr, migration_id,
+    );
+    info!(log, "Begin migration"; "src_migrate_url" => &src_migrate_url);
+    let (mut conn, _) =
+        tokio_tungstenite::connect_async(src_migrate_url).await?;
+
+    let dst_protocols = super::protocol::make_protocol_offer();
+    conn.send(tungstenite::Message::Text(dst_protocols)).await?;
+    let selected = match conn.next().await {
+        Some(Ok(tungstenite::Message::Text(selected_protocol))) => {
+            info!(log, "source negotiated protocol {}", selected_protocol);
+            match super::protocol::select_protocol_from_offer(
+                &selected_protocol,
+            ) {
+                Ok(Some(selected)) => selected,
+                Ok(None) => {
+                    let offered = super::protocol::make_protocol_offer();
+                    error!(log, "source selected protocol not on offer";
+                           "offered" => &offered,
+                           "selected" => &selected_protocol);
+
+                    return Err(MigrateError::NoMatchingProtocol(
+                        selected_protocol,
+                        offered,
+                    ));
+                }
+                Err(e) => {
+                    error!(log, "source selected protocol failed to parse";
+                           "selected" => &selected_protocol);
+
+                    return Err(MigrateError::ProtocolParse(
+                        selected_protocol,
+                        e.to_string(),
+                    ));
+                }
+            }
+        }
+        x => {
+            conn.send(tungstenite::Message::Close(Some(CloseFrame {
+                code: CloseCode::Protocol,
+                reason: "did not respond to version handshake.".into(),
+            })))
+            .await?;
+            error!(
+                log,
+                "source instance failed to negotiate protocol version: {:?}", x
+            );
+            return Err(MigrateError::Initiate);
+        }
     };
 
-    if let Err(err) = proto.run().await {
-        let _ = err_tx
-            .send(MigrateTargetCommand::UpdateState(MigrationState::Error))
-            .await;
-
-        // We encountered an error, try to inform the remote before bailing
-        // Note, we don't use `?` here as this is a best effort and we don't
-        // want an error encountered during this send to shadow the run error
-        // from the caller.
-        if let Ok(e) = codec::Message::Error(err.clone()).try_into() {
-            let _ = proto.conn.send(e).await;
+    Ok(match selected {
+        Protocol::RonV0 => RonV0::new(log, migration_id, conn, local_addr),
+    })
+}
+
+/// A helper type for abstracting away type-level differences between the
+/// different phases of instance ensure.
+enum EnsureState<'a> {
```

Comment:
This wrapper enum and its usage below are the other major source of net-new code in this PR.